File: impl/typescript/src/account/validation.ts
/**
 * Account validation -- type-specific rule enforcement.
 *
 * Spec reference: 001 spec.md, data-model.md
 *   - FR-006: Parent validation rules V-PARENT-01..05.
 *   - FR-007: Child validation rules V-CHILD-01..05.
 *   - FR-007a: Shared validation rules V-SHARED-01..04.
 *   - FR-014: Structured error reporting with all violations collected.
 *   - SC-005: All violations reported, not just the first.
 *
 * The validateAccount function checks all applicable rules for the given
 * account type and returns an array of ALL validation errors found.
 * An empty array indicates a valid account.
 *
 * Note: Several rules (V-PARENT-04, V-PARENT-05, V-CHILD-05, V-SHARED-02,
 * V-SHARED-03, V-SHARED-04) are enforced at compile time by the TypeScript
 * type system (discriminated union on accountType). These rules cannot be
 * violated at runtime if the account was constructed via the builder.
 * The runtime validation focuses on rules that could be violated if
 * accounts are constructed directly (e.g., from deserialization).
 */

import type { NeuronAccount } from './account.js';
import { isParent, isChild, isShared } from './account.js';
import {
  ACCT_MISSING_REQUIRED_FIELD,
  ACCT_INVALID_ACCOUNT_TYPE,
} from '../errors.js';

/**
 * A single validation error describing a rule violation.
 *
 * FR-014: Structured error reporting with field name, rule ID, and message.
 */
export interface ValidationError {
  /** The field that caused the validation failure. */
  readonly field: string;
  /** The validation rule ID (e.g., "V-PARENT-01"). */
  readonly rule: string;
  /** The error code from the NEURON-ACCT-* taxonomy. */
  readonly code: string;
  /** Human-readable description of the violation. */
  readonly message: string;
}

/**
 * Validate a NeuronAccount against its type-specific rules.
 *
 * FR-014, SC-005: Collects ALL violations -- does not stop at the first error.
 * Returns an empty array for a structurally valid account.
 *
 * Rules checked at runtime:
 * - V-PARENT-01: Parent must have DID (NEURON-ACCT-002)
 * - V-PARENT-02: Parent must have publicKey (NEURON-ACCT-002)
 * - V-CHILD-01: Child must have parentPubKey (NEURON-ACCT-002)
 * - V-CHILD-02: Child must have publicKey (NEURON-ACCT-002)
 * - V-SHARED-01: Shared must have multisigKey (NEURON-ACCT-002)
 *
 * Rules enforced by the type system (not checked at runtime):
 * - V-PARENT-03: creditBalance capability (field exists on ParentAccount)
 * - V-PARENT-04: No parentPubKey (not on ParentAccount interface)
 * - V-PARENT-05: No multisigKey (not on ParentAccount interface)
 * - V-CHILD-04: DID is optional (no error for absence)
 * - V-CHILD-05: No multisigKey (not on ChildAccount interface)
 * - V-SHARED-02: No DID (not on SharedAccount interface)
 * - V-SHARED-03: No parentPubKey (not on SharedAccount interface)
 * - V-SHARED-04: No publicKey (not on SharedAccount interface)
 *
 * @param account - NeuronAccount to validate
 * @returns Array of validation errors (empty if valid)
 */
export function validateAccount(account: NeuronAccount): ValidationError[] {
  const errors: ValidationError[] = [];

  if (isParent(account)) {
    validateParent(account, errors);
  } else if (isChild(account)) {
    validateChild(account, errors);
  } else if (isShared(account)) {
    validateShared(account, errors);
  } else {
    // Exhaustive check: should never reach here with valid discriminated union
    errors.push({
      field: 'accountType',
      rule: 'V-TYPE-01',
      code: ACCT_INVALID_ACCOUNT_TYPE,
      message: `Unknown account type: ${String((account as NeuronAccount).accountType)}`,
    });
  }

  return errors;
}

/**
 * Validate Parent-specific rules.
 *
 * V-PARENT-01: Must have DID.
 * V-PARENT-02: Must have publicKey.
 */
function validateParent(
  account: { readonly did?: unknown; readonly publicKey?: unknown },
  errors: ValidationError[],
): void {
  // V-PARENT-01: Must have DID
  if (account.did == null) {
    errors.push({
      field: 'did',
      rule: 'V-PARENT-01',
      code: ACCT_MISSING_REQUIRED_FIELD,
      message: 'Parent account must have a DID',
    });
  }

  // V-PARENT-02: Must have publicKey
  if (account.publicKey == null) {
    errors.push({
      field: 'publicKey',
      rule: 'V-PARENT-02',
      code: ACCT_MISSING_REQUIRED_FIELD,
      message: 'Parent account must have a publicKey',
    });
  }
}

/**
 * Validate Child-specific rules.
 *
 * V-CHILD-01: Must have parentPubKey.
 * V-CHILD-02: Must have publicKey.
 * V-CHILD-04: DID is optional (not an error if absent).
 */
function validateChild(
  account: { readonly parentPubKey?: unknown; readonly publicKey?: unknown },
  errors: ValidationError[],
): void {
  // V-CHILD-01: Must have parentPubKey
  if (account.parentPubKey == null) {
    errors.push({
      field: 'parentPubKey',
      rule: 'V-CHILD-01',
      code: ACCT_MISSING_REQUIRED_FIELD,
      message: 'Child account must have a parentPubKey',
    });
  }

  // V-CHILD-02: Must have publicKey
  if (account.publicKey == null) {
    errors.push({
      field: 'publicKey',
      rule: 'V-CHILD-02',
      code: ACCT_MISSING_REQUIRED_FIELD,
      message: 'Child account must have a publicKey',
    });
  }
}

/**
 * Validate Shared-specific rules.
 *
 * V-SHARED-01: Must have multisigKey.
 */
function validateShared(
  account: { readonly multisigKey?: unknown },
  errors: ValidationError[],
): void {
  // V-SHARED-01: Must have multisigKey
  if (account.multisigKey == null) {
    errors.push({
      field: 'multisigKey',
      rule: 'V-SHARED-01',
      code: ACCT_MISSING_REQUIRED_FIELD,
      message: 'Shared account must have a multisigKey',
    });
  }
}

File: impl/golang/internal/account/validation.go
package account

import "fmt"

// FR-014: Actionable error messages on validation failure
// ValidationError represents a structured validation failure for account construction
// or mutation. Each error pinpoints the offending field, a machine-readable rule code,
// and a human-readable actionable message.
//
// Rule codes follow the pattern "V-{TYPE}-{NN}" where TYPE is PARENT, CHILD, or SHARED,
// and NN is a zero-padded sequence number (e.g. "V-PARENT-01", "V-CHILD-03").
type ValidationError struct {
	Field    string // the field that failed validation, e.g. "publicKey", "did"
	RuleCode string // machine-readable rule code, e.g. "V-PARENT-01"
	Message  string // human-readable, actionable description of the failure
}

// Error implements the error interface, returning a formatted string that includes
// the rule code, field name, and message in the form "[RuleCode] Field: Message".
func (e ValidationError) Error() string {
	return fmt.Sprintf("[%s] %s: %s", e.RuleCode, e.Field, e.Message)
}

// Validate checks ALL validation rules for this account's type and returns every
// violation found. It does not short-circuit on the first error; callers receive
// the complete list of problems so they can be displayed or logged together.
//
// Returns an empty slice when the account satisfies all rules for its type.
// Returns a single-element slice with a generic error for Unspecified (zero-value)
// account types.
func (a *NeuronAccount) Validate() []ValidationError {
	switch a.accountType {
	case Parent:
		return a.validateParent()
	case Child:
		return a.validateChild()
	case Shared:
		return a.validateShared()
	default:
		return []ValidationError{
			{
				Field:    "accountType",
				RuleCode: "V-TYPE-00",
				Message:  "account type is unspecified or unknown; must be Parent, Child, or Shared",
			},
		}
	}
}

// validateParent checks rules V-PARENT-01 through V-PARENT-05.
func (a *NeuronAccount) validateParent() []ValidationError {
	var errs []ValidationError

	// V-PARENT-01: Must have DID (did != nil && did.identifier != "")
	if a.did == nil || a.did.identifier == "" {
		errs = append(errs, ValidationError{
			Field:    "did",
			RuleCode: "V-PARENT-01",
			Message:  "Parent account must have a DID",
		})
	}

	// V-PARENT-02: Must have single NeuronPublicKey (publicKey != nil && !isZeroPublicKey)
	if a.publicKey == nil || isZeroPublicKey(*a.publicKey) {
		errs = append(errs, ValidationError{
			Field:    "publicKey",
			RuleCode: "V-PARENT-02",
			Message:  "Parent account must have a single NeuronPublicKey",
		})
	}

	// V-PARENT-03: Must have currency symbol (currencySymbol != "")
	if a.currencySymbol == "" {
		errs = append(errs, ValidationError{
			Field:    "currencySymbol",
			RuleCode: "V-PARENT-03",
			Message:  "Parent account must have a currency symbol",
		})
	}

	// V-PARENT-04: Must NOT have parent reference (parentPubKey == nil)
	if a.parentPubKey != nil {
		errs = append(errs, ValidationError{
			Field:    "parentPubKey",
			RuleCode: "V-PARENT-04",
			Message:  "Parent account must not have a parent public key reference",
		})
	}

	// V-PARENT-05: Must NOT have multisig key (multisigKey == nil)
	if a.multisigKey != nil {
		errs = append(errs, ValidationError{
			Field:    "multisigKey",
			RuleCode: "V-PARENT-05",
			Message:  "Parent account must not have a multisig key",
		})
	}

	return errs
}

// validateChild checks rules V-CHILD-01 through V-CHILD-05.
func (a *NeuronAccount) validateChild() []ValidationError {
	var errs []ValidationError

	// V-CHILD-01: Must have parent NeuronPublicKey ref (parentPubKey != nil && !isZeroPublicKey)
	if a.parentPubKey == nil || isZeroPublicKey(*a.parentPubKey) {
		errs = append(errs, ValidationError{
			Field:    "parentPubKey",
			RuleCode: "V-CHILD-01",
			Message:  "Child account must reference a parent NeuronPublicKey",
		})
	}

	// V-CHILD-02: Must have single NeuronPublicKey (publicKey != nil && !isZeroPublicKey)
	if a.publicKey == nil || isZeroPublicKey(*a.publicKey) {
		errs = append(errs, ValidationError{
			Field:    "publicKey",
			RuleCode: "V-CHILD-02",
			Message:  "Child account must have a single NeuronPublicKey",
		})
	}

	// V-CHILD-03: Must have registry binding (registryBinding != nil)
	if a.registryBinding == nil {
		errs = append(errs, ValidationError{
			Field:    "registryBinding",
			RuleCode: "V-CHILD-03",
			Message:  "Child account must have a registry binding",
		})
	}

	// V-CHILD-04: Must NOT have DID (did == nil)
	if a.did != nil {
		errs = append(errs, ValidationError{
			Field:    "did",
			RuleCode: "V-CHILD-04",
			Message:  "Child account must not have a DID",
		})
	}

	// V-CHILD-05: Must NOT have multisig key (multisigKey == nil)
	if a.multisigKey != nil {
		errs = append(errs, ValidationError{
			Field:    "multisigKey",
			RuleCode: "V-CHILD-05",
			Message:  "Child account must not have a multisig key",
		})
	}

	return errs
}

// validateShared checks rules V-SHARED-01 through V-SHARED-04.
func (a *NeuronAccount) validateShared() []ValidationError {
	var errs []ValidationError

	// V-SHARED-01: Must have MultisigKey with threshold (multisigKey != nil)
	if a.multisigKey == nil {
		errs = append(errs, ValidationError{
			Field:    "multisigKey",
			RuleCode: "V-SHARED-01",
			Message:  "Shared account must have a MultisigKey with threshold",
		})
	}

	// V-SHARED-02: Must NOT have DID (did == nil)
	if a.did != nil {
		errs = append(errs, ValidationError{
			Field:    "did",
			RuleCode: "V-SHARED-02",
			Message:  "Shared account must not have a DID",
		})
	}

	// V-SHARED-03: Must NOT have parent reference (parentPubKey == nil)
	if a.parentPubKey != nil {
		errs = append(errs, ValidationError{
			Field:    "parentPubKey",
			RuleCode: "V-SHARED-03",
			Message:  "Shared account must not have a parent public key reference",
		})
	}

	// V-SHARED-04: Must NOT have single publicKey (publicKey == nil)
	if a.publicKey != nil {
		errs = append(errs, ValidationError{
			Field:    "publicKey",
			RuleCode: "V-SHARED-04",
			Message:  "Shared account must not have a single public key",
		})
	}

	return errs
}
